### ISSUE TITLE
How to normalize a bipartite layer

### ISSUE THREAD

**Issue by sasan73** [rusty1s/pytorch_geometric#2048]
## ❓ Questions & Help
Hi, Thank you for your hard work. 
My code looks like this
```python 
class GCMCLayer(MessagePassing): 
  def __init__(self, in_channel1, in_channel2):
    super(GCMCLayer, self).__init__(aggr = 'add')
    self.in_channel1 = in_channel1
    self.in_channel2 = in_channel2
    self.lin1 = nn.Linear(self.in_channel1, 12)
    self.lin2 = nn.Linear(self.in_channel2, 12)
  def forward(self, x, edge_index, N, M):

    if isinstance(x, tuple):
      x1, x2 = x
    else:
      raise Exception(f"x requires a tuple not a {type(x)}") 

    if x1.shape[1] == self.in_channel1:
      x1 = self.lin1(x1)
      x2 = self.lin2(x2)
    elif x1.shape[1] == self.in_channel2:
      x1 = self.lin2(x1)
      x2 = self.lin1(x2)
    else:
      raise Exception("input channels do not match the number of nodes")

    row, col = edge_index

    deg_i = degree(col)
    deg_u = degree(row)
    deg_inv_i = deg_i.pow(-0.5)
    deg_inv_u = deg_u.pow(-0.5)
    norm = deg_inv_u[row] * deg_inv_i[col]


    return self.propagate(edge_index, x=x, norm = norm, size = (N, M))

    def message(self, x_j, norm):
      return norm.view(-1, 1) * x_j
```
I gave it some inputs of the data that I has as such: 
```python
conv = GCMCLayer(2, 3)

out1 = conv((torch.cat((train.x_u, torch.ones(size=(train.x_u.shape[0],)).unsqueeze(1)), dim = 1), train.x_i), train.edge_index, train.x_u.shape[0], train.x_i.shape[0])
out2 = conv((train.x_i, torch.cat((train.x_u, torch.ones(size=(train.x_u.shape[0],)).unsqueeze(1)), dim = 1)), train.edge_index[torch.tensor([1,0])], train.x_i.shape[0], train.x_u.shape[0])
```
And the output looks like this
```python 
tensor([[6.8845e+04, 5.1800e+02, 9.0000e+00],
        [1.5852e+06, 1.3666e+05, 2.1500e+02],
        [2.4531e+06, 7.1958e+04, 3.4900e+02],
        ...,
        [4.3335e+06, 6.0196e+05, 6.1700e+02],
        [4.7698e+06, 9.4300e+02, 5.7300e+02],
        [1.5694e+06, 4.1846e+04, 2.1500e+02]])
```
```python
tensor([[   4.,   85.],
        [ 136., 3257.],
        [  37., 1258.],
        ...,
        [   6.,  149.],
        [   6.,  149.],
        [   6.,  149.]])
```
My question is why are the outputs **integers**? should they not have been Normalized? 
I do have another question when I want to use a linear function on ```x_j``` in ```message``` I can use a ```nn.Linear``` of arbitrary shape and won't get any error regarding the dimensionality of x_j. I'm having a hard time understanding how x_j works. I would really appreciate your clarification.

**Comment by rusty1s** [comment]
How does your data `train` look like? Note that you do not make use of transformed input features `x1` and `x2` during propagation. You may want to change your code to this:
```
return self.propagate(edge_index, x=(x1, x2), norm = norm, size = (N, M))
```
